File: tests/jnet/TestHeader.java
import java.util.Date;
import java.util.ArrayList;
import org.junit.Test;
import org.junit.Assert;
import org.junit.Assume;
import jnet.Bytes;
import jnet.CRC;
import jnet.Header;
import jnet.MalformedDataException;


public class TestHeader {

    @Test
    public void testAttachParse() {
        for (int i = 0; i < TestMain.RANDOM_TEST_AMOUNT; i++) {
            int dataLength = TestMain.RANDOM_DIS.nextInt(1000);
            String data = TestMain.randomString(dataLength);
            byte[] payload = Bytes.serialize(data);
            byte[] body = CRC.attach(payload);

            Assume.assumeTrue(CRC.check(body));

            byte[] message = Header.attach(body);
            byte[] header = new byte[Header.SIZE];
            System.arraycopy(message, 0, header, 0, header.length);
            Header.Info headerInfo = Header.validateAndParse(header);

            Assert.assertNotNull(headerInfo);
            Assert.assertEquals(headerInfo.id, Header.HEADER_BYTE);
            Assert.assertEquals(headerInfo.size, body.length);
        }
    }

    @Test
    public void testAttachToNullArgument() {
        Assert.assertThrows(NullPointerException.class, () -> Header.attach(null));
    }

    @Test
    public void testValidateWithNullArgument() {
        Assert.assertThrows(NullPointerException.class, () -> Header.attach(null));
    }

    @Test
    public void testMissingCrc() {
        String data = TestMain.randomString(100);
        byte[] payload = Bytes.serialize(data);

        Assert.assertThrows(MalformedDataException.class, () -> Header.attach(payload));
    }

    @Test
    public void testInvalidCrc() {
        String data = TestMain.randomString(100);
        String dataForCrc = TestMain.randomString(1);
        byte[] payload = Bytes.serialize(data);
        byte[] payloadForCrc = Bytes.serialize(dataForCrc);

        byte[] crc = Bytes.intToBytes(CRC.crc32(payloadForCrc));
        byte[] body = new byte[payload.length + crc.length];
        System.arraycopy(payload, 0, body, 0, payload.length);
        System.arraycopy(crc, 0, body, body.length - crc.length, crc.length);

        Assert.assertThrows(MalformedDataException.class, () -> Header.attach(payload));
    }

}

File: tests/pnet/test_header.py
import random
import pytest
import conftest
import struct
from pnet import byteutils
from pnet import crc
from pnet import header
from pnet.error import MalformedDataError


class TestHeader:

    def test_attach_parse(self):
        for i in range(conftest.RANDOM_TEST_AMOUNT):
            data_len: int = random.randint(0, 1000)
            data: str = conftest.random_str(data_len)
            payload: bytes = byteutils.str_to_bytes(data)
            body: bytes = crc.attach(payload)

            assert crc.check(body)

            message: bytes = header.attach(body)
            header_bytes: bytes = message[0:header.SIZE]
            header_info: header.Info = header.validate_and_parse(header_bytes)

            assert header_info is not None
            assert header_info.id == header.HEADER_BYTE
            assert header_info.size == len(body)

    def test_attach_to_null_argument(self):
        with pytest.raises(TypeError):
            header.attach(None)

    def test_validate_with_null_argument(self):
        with pytest.raises(TypeError):
            header.validate_and_parse(None)

    def test_missing_crc(self):
        data: str = conftest.random_str(100)
        payload: bytes = byteutils.str_to_bytes(data)

        with pytest.raises(MalformedDataError):
            header.attach(payload)

    def test_invalid_crc(self):
        data: str = conftest.random_str(100)
        data_for_crc: str = conftest.random_str(1)
        payload: bytes = byteutils.str_to_bytes(data)
        payload_for_crc: bytes = byteutils.str_to_bytes(data_for_crc)

        crc_bytes: bytes = struct.pack("<I", crc.crc32(payload_for_crc))
        body: bytes = payload + crc_bytes

        with pytest.raises(MalformedDataError):
            header.attach(body)
